### ISSUE TITLE
Add admin view to access add-on files regardless of status

### ISSUE THREAD

**Issue by wagnerand** [mozilla/addons-server#8985]
For security reasons, admins need to have access to any and all add-on files on AMO. Currently, the reviewer tools don't support accessing files for deleted add-ons.

@diox suggested adding an admin view that lets admins access files for deleted add-ons. I'd like to extend this to: access files regardless of their status or the status of the add-on, so admins have a single place where they can access files they need.

This should be gated by an appropriate permission, maybe @jvillalobos can weigh in whether we can use an existing one or need a new one.

**Comment by diox** [comment]
Untested, but something like this should work:
```python
## in zadmin/urls.py:

url('^download_any_file/(?P<file_id>\d+)', views.download_any_file,
    name='zadmin.download_any_file'),


## in zadmin/views.py:

# @admin_required will require Admin:Tools, which staff has. We could change it
# to something more restrictive like Admin:Advanced.
@admin_required
def download_any_file(request, file_id)
    file_ = get_object_or_404(File.objects, pk=file_id)
    addon = get_object_or_404(Addon.unfiltered,
                              pk=file_.version.addon_id)
    return download_file(request, file_id, file_=file_, addon=addon)


# Note: zadmin already has a download_file view, but confusingly it's to
# download fileuploads. Consider renaming the view while we're at it. Maybe
# even the url.
```

**Comment by muffinresearch** [comment]
@wagnerand I've assigned this to the backlog - are you OK with that or is the need for this more urgent?

**Comment by wagnerand** [comment]
There is an urgent need for some particular files (which I have requested in [bug 1478346](https://bugzilla.mozilla.org/show_bug.cgi?id=1478346)) for [bug 1475330](https://bugzilla.mozilla.org/show_bug.cgi?id=1475330) . Generally, we won't need this on a daily basis, but whenever we do, it will be urgent. The priority of implementing this will depend on how often and how fast ops can provide those files on demand).

**Comment by wagnerand** [comment]
@muffinresearch Digging out the files manually from the production filesystem turned out to be significant effort for anyone who does it, and very few people actually have access to do this. Therefore, I am raising the priority to P2.

/cc @bqbn who did it for me this time in the aforementioned bug.

**Comment by EnTeQuAk** [comment]
QA and everyone else who has to use it:

* The file admin (/admin/models/files/) has been improved as part of this
 * There is a bit more information in the file-list and you can search by slug and guid (not a full featured search but "starts with" search so that exact matches are probably ideal) of an add-on
* The detail page has a bit more structured information about a file
* As mentioned in the PR screenshot, there is now a download link for each file

@wagnerand let me know if this works fine for you or if we should improve this over time in case you need more or different stuff.

**Comment by AlexandraMoga** [comment]
I've verified this new functionalities on stage. I'm not sure why, but -dev is still showing the old version of `/admin/models/files/`

Here are my observations:

**Search**
1. files can be found when searching by an add-on slug or guid (exact match or nearly exact)
2.  _exception_: if an add-on is deleted, the slug is removed, so searching will only be possible by guid

**File download:**
1. a Download File link is present in the file details page. When clicked, an install is triggered. To download the file, the user needs to use right click -> Save Link As.. option.  I think it would be better if the link triggered the download process directly rather than an installation at click. 
2.  downloads are working  for all 'Approved' add-on version files
3.  for add-ons with the following status: Invisible, Disabled by user or by Mozilla -> I could download the file but the archive can't be opened. The file does look like an `.xpi`, but if I test the package it says that it is not an archive... Try downloading this file - https://addons-internal.stage.mozaws.net/en-US/admin/models/files/file/325747/change/?_changelist_filters=q%3Dsource-code-upload
4. for deleted add-ons - in the file details, the status is still set to 'Approved' -> files can be downloaded and inspected. I'm not sure if  we should display an 'Approved' status though if the add-on version was deleted - see https://addons-internal.stage.mozaws.net/en-US/admin/models/files/file/322499/change/?_changelist_filters=q%3D%257Ba68b25ee-16ba-4827-8ae6-a6f9048fe605%257D
5. for Unlisted add-ons - even if the files are Approved, I have the same behavior described at point 3. Not sure if Unlisted add-ons are in scope though. 

**another issue**
- the link next to the Version field that should lead to `https://addons-internal.stage.mozaws.net/en-US/admin/models/versions/version/<number>/change/` is triggering a 504 error

@EnTeQuAk @wagnerand please have  a look at these results and let me know what you think

**Comment by wagnerand** [comment]
> @wagnerand let me know if this works fine for you or if we should improve this over time in case you need more or different stuff.

@EnTeQuAk I will give it a shot next time I need this, thank you!

> a Download File link is present in the file details page. When clicked, an install is triggered. To download the file, the user needs to use right click -> Save Link As.. option. I think it would be better if the link triggered the download process directly rather than an installation at click.

Yes, I would very much prefer that (the same request is filed for the links in the reviewer tools)

> for add-ons with the following status: Invisible, Disabled by user or by Mozilla -> I could download the file but the archive can't be opened. The file does look like an .xpi, but if I test the package it says that it is not an archive... Try downloading this file - addons-internal.stage.mozaws.net/en-US/admin/models/files/file/325747/change/?_changelist_filters=q%3Dsource-code-upload

That sounds very much like a bug to me because that file link is actually 404. @AlexandraMoga you can test that by left-clicking the link.

> for deleted add-ons - in the file details, the status is still set to 'Approved' -> files can be downloaded and inspected. I'm not sure if we should display an 'Approved' status though if the add-on version was deleted - see addons-internal.stage.mozaws.net/en-US/admin/models/files/file/322499/change/?_changelist_filters=q%3D%257Ba68b25ee-16ba-4827-8ae6-a6f9048fe605%257D

Yeah, we don't have a file status for a deleted version at the moment. I think it would make sense to set the file status to `STATUS_DISABLED` when a version is deleted.

> for Unlisted add-ons - even if the files are Approved, I have the same behavior described at point 3. Not sure if Unlisted add-ons are in scope though.

This should work for all add-ons. I suppose it's the same bug as 3, so fixing that will likely fix this as well.

> the link next to the Version field that should lead to https://addons-internal.stage.mozaws.net/en-US/admin/models/versions/version/<number>/change/ is triggering a 504 error

Yeah, I can reproduce that, getting `504 Gateway Time-out`.

@AlexandraMoga Would ypu mind filing a follow-up issue for
* making the file link work (item 3 and 5)
* making the file link trigger a download instead of an install (item 1)
* the 504 error

@EnTeQuAk what do you think about setting the file status to `STATUS_DISABLED` for deleted versions?
